File: benchmarks/json_parse/json_parse.py
SRC = '{"users":[{"id":1,"name":"Alice","active":true,"scores":[85,92,78]},{"id":2,"name":"Bob","active":false,"scores":[91,88,95]},{"id":3,"name":"Charlie","active":true,"scores":[76,84,90]}],"metadata":{"version":2,"generated":false,"tags":["benchmark","json","test"],"config":{"timeout":30,"retries":3}}}'


class Parser:
    def __init__(self, src):
        self.src = src
        self.pos = 0

    def peek(self):
        return ord(self.src[self.pos]) if self.pos < len(self.src) else 0

    def advance(self):
        self.pos += 1

    def eof(self):
        return self.pos >= len(self.src)

    def skip_ws(self):
        while not self.eof() and chr(self.peek()) in ' \t\n\r':
            self.advance()

    @staticmethod
    def is_digit(c):
        return 48 <= c <= 57

    def parse_value(self):
        self.skip_ws()
        if self.eof():
            return 0
        c = self.peek()
        if c == 34:   # '"'
            return self.parse_string()
        if c == 116:  # 't'
            self.advance(); self.advance(); self.advance(); self.advance()
            return 1
        if c == 102:  # 'f'
            self.advance(); self.advance(); self.advance(); self.advance(); self.advance()
            return 0
        if c == 110:  # 'n'
            self.advance(); self.advance(); self.advance(); self.advance()
            return 0
        if c == 45 or self.is_digit(c):  # '-'
            return self.parse_number()
        if c == 91:   # '['
            return self.parse_array()
        if c == 123:  # '{'
            return self.parse_object()
        return 0

    def parse_string(self):
        self.advance()
        n = 0
        while self.peek() != 34:
            self.advance()
            n += 1
        self.advance()
        return n

    def parse_number(self):
        sign = 1
        if self.peek() == 45:
            self.advance()
            sign = -1
        val = 0
        while not self.eof() and self.is_digit(self.peek()):
            val = val * 10 + (self.peek() - 48)
            self.advance()
        return sign * val

    def parse_array(self):
        self.advance()
        self.skip_ws()
        if self.peek() == 93:
            self.advance()
            return 0
        s = 0
        while True:
            s += self.parse_value()
            self.skip_ws()
            if self.peek() == 44:
                self.advance()
            else:
                break
        self.advance()
        return s

    def parse_object(self):
        self.advance()
        self.skip_ws()
        if self.peek() == 125:
            self.advance()
            return 0
        s = 0
        while True:
            self.skip_ws()
            self.parse_string()
            self.skip_ws()
            self.advance()  # ':'
            s += self.parse_value()
            self.skip_ws()
            if self.peek() == 44:
                self.advance()
            else:
                break
        self.advance()
        return s


if __name__ == "__main__":
    parser = Parser(SRC)
    total = 0
    for _ in range(5000):
        parser.pos = 0
        total += parser.parse_value()
    print(total)

File: benchmarks/json_parse/json_parse.rs
const SRC: &str = r#"{"users":[{"id":1,"name":"Alice","active":true,"scores":[85,92,78]},{"id":2,"name":"Bob","active":false,"scores":[91,88,95]},{"id":3,"name":"Charlie","active":true,"scores":[76,84,90]}],"metadata":{"version":2,"generated":false,"tags":["benchmark","json","test"],"config":{"timeout":30,"retries":3}}}"#;

struct Parser<'a> {
    src: &'a [u8],
    pos: usize,
}

impl<'a> Parser<'a> {
    fn peek(&self) -> u8 { self.src[self.pos] }
    fn advance(&mut self) { self.pos += 1; }
    fn eof(&self) -> bool { self.pos >= self.src.len() }

    fn skip_ws(&mut self) {
        while !self.eof() {
            let c = self.peek();
            if c == b' ' || c == b'\t' || c == b'\n' || c == b'\r' { self.advance(); }
            else { break; }
        }
    }

    fn is_digit(c: u8) -> bool { c >= b'0' && c <= b'9' }

    fn parse_value(&mut self) -> i64 {
        self.skip_ws();
        if self.eof() { return 0; }
        match self.peek() {
            b'"' => self.parse_string(),
            b't' => { self.advance(); self.advance(); self.advance(); self.advance(); 1 }
            b'f' => { self.advance(); self.advance(); self.advance(); self.advance(); self.advance(); 0 }
            b'n' => { self.advance(); self.advance(); self.advance(); self.advance(); 0 }
            b'-' | b'0'..=b'9' => self.parse_number(),
            b'[' => self.parse_array(),
            b'{' => self.parse_object(),
            _ => 0,
        }
    }

    fn parse_string(&mut self) -> i64 {
        self.advance();
        let mut n: i64 = 0;
        while self.peek() != b'"' {
            self.advance();
            n += 1;
        }
        self.advance();
        n
    }

    fn parse_number(&mut self) -> i64 {
        let mut sign: i64 = 1;
        if self.peek() == b'-' { self.advance(); sign = -1; }
        let mut val: i64 = 0;
        while !self.eof() && Self::is_digit(self.peek()) {
            val = val * 10 + ((self.peek() - b'0') as i64);
            self.advance();
        }
        sign * val
    }

    fn parse_array(&mut self) -> i64 {
        self.advance(); self.skip_ws();
        if self.peek() == b']' { self.advance(); return 0; }
        let mut sum: i64 = 0;
        loop {
            sum += self.parse_value();
            self.skip_ws();
            if self.peek() == b',' { self.advance(); } else { break; }
        }
        self.advance();
        sum
    }

    fn parse_object(&mut self) -> i64 {
        self.advance(); self.skip_ws();
        if self.peek() == b'}' { self.advance(); return 0; }
        let mut sum: i64 = 0;
        loop {
            self.skip_ws();
            self.parse_string();  // discard key
            self.skip_ws();
            self.advance();  // ':'
            sum += self.parse_value();
            self.skip_ws();
            if self.peek() == b',' { self.advance(); } else { break; }
        }
        self.advance();
        sum
    }
}

fn main() {
    let src_bytes = SRC.as_bytes();
    let mut parser = Parser { src: src_bytes, pos: 0 };
    let mut total: i64 = 0;
    for _ in 0..100_000 {
        parser.pos = 0;
        total += parser.parse_value();
    }
    println!("{}", total);
}
